src/NuGet.Core/NuGet.Common/KeyedMutex.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Common
+{
+    internal sealed class KeyedMutex : IDisposable
+    {
+        private readonly Dictionary<string, LockState> _locks;
+        private readonly SemaphoreSlim _dictionaryLock;
+
+        internal KeyedMutex()
+        {
+            _locks = new Dictionary<string, LockState>();
+            _dictionaryLock = new SemaphoreSlim(initialCount: 1);
+        }
+
+        internal async Task EnterAsync(string key, CancellationToken token)
```

Comment:
Does this have the same case-sensitivity properties as the previous implementation? If file-only approach ran on Windows it would "lock the string" in a case insensitive manner, correct? `_locks` seems to be case sensitive. I see that the input string in the calling class is ToUpper'd then hashed so I'm not sure what the expectation is from the caller or what the implementation is supposed to do.

Reply:
Looking at `ConcurrencyUtilities`, the original path goes through `ToUpper`, then a SHA1 hash, to generate a filename used for locking. If the input to `ConcurrentUtilities` is called twice with different casing, and the `ToUpper` wasn't there (or has bugs like the Turkish-i issue), then the SHA1 hash would be very different, and the filesystem lock wouldn't work anyway. Anyway, since the filename used for locking is always the output of the SHA1 hash, we can be confident that it's deterministic, so the dictionary used doesn't need to be case-insensitive.